pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -681,182 +700,172 @@ func buildSidecarOutboundHTTPListenerOptsForPortOrUDS(listenerMapKey *string,
-			svcListenAddress == constants.UnspecifiedIP {
-			svcListenAddress = constants.UnspecifiedIPv6
-		}
-		// We should never get an empty address.
-		// This is a safety guard, in case some platform adapter isn't doing things
-		// properly
-		if len(svcListenAddress) > 0 {
-			if !strings.Contains(svcListenAddress, "/") {
-				listenerOpts.bind = svcListenAddress
-			} else {
-				// Address is a CIDR. Fall back to 0.0.0.0 and
-				// filter chain match
-				destinationCIDR = svcListenAddress
-				listenerOpts.bind = actualWildcard
+func buildSidecarOutboundTCPListenerOpts(opts outboundListenerOpts, virtualServices []config.Config) []*filterChainOpts {
+	meshGateway := map[string]bool{constants.IstioMeshGateway: true}
+	out := make([]*filterChainOpts, 0)
+	var svcConfigs []config.Config
+	if opts.service != nil {
+		// Do not filter namespace for now.
+		// TODO(https://github.com/istio/istio/issues/46146) we may need to, or something more sophisticated
+		svcConfigs = getConfigsForHost("", opts.service.Hostname, virtualServices)
+	} else {
+		svcConfigs = virtualServices
+	}
+
+	out = append(out, buildSidecarOutboundTLSFilterChainOpts(opts.proxy, opts.push, opts.cidr, opts.service,
+		opts.bind.Primary(), opts.port, meshGateway, svcConfigs)...)
+	out = append(out, buildSidecarOutboundTCPFilterChainOpts(opts.proxy, opts.push, opts.cidr, opts.service,
+		opts.port, meshGateway, svcConfigs)...)
+	return out
+}
+
+// buildSidecarOutboundListener builds a single listener and
+// adds it to the listenerMap provided by the caller.  Listeners are added
+// if one doesn't already exist. HTTP listeners on same port are ignored
+// (as vhosts are shipped through RDS).  TCP listeners on same port are
+// allowed only if they have different CIDR matches.
+func (lb *ListenerBuilder) buildSidecarOutboundListener(listenerOpts outboundListenerOpts,
+	listenerMap map[string]*outboundListenerEntry, virtualServices []config.Config, actualWildcards []string,
+) {
+	// TODO: remove actualWildcard
+	var currentListenerEntry *outboundListenerEntry
+
+	conflictType := NoConflict
+
+	listenerPortProtocol := listenerOpts.port.Protocol
+	listenerProtocol := istionetworking.ModelProtocolToListenerProtocol(listenerOpts.port.Protocol)
+
+	var listenerMapKey string
+	switch listenerProtocol {
+	case istionetworking.ListenerProtocolTCP, istionetworking.ListenerProtocolAuto:
+		// Determine the listener address if bind is empty
+		// we listen on the service VIP if and only
+		// if the address is an IP address. If its a CIDR, we listen on
+		// 0.0.0.0, and setup a filter chain match for the CIDR range.
+		// As a small optimization, CIDRs with /32 prefix will be converted
+		// into listener address so that there is a dedicated listener for this
+		// ip:port. This will reduce the impact of a listener reload
+		if len(listenerOpts.bind.Primary()) == 0 { // TODO: make this better
```

Comment:
nit: comparing checking `== ""` would be clearer here. Unless you mean to check `len(listenerOpts.binds)`?